### DISCUSSION TITLE
Correlating Requests and Responses in c-ares

### DISCUSSION THREAD

**Discussion by arnnav** [c-ares/c-ares#928] [Q&A]
Hello c-ares community,

I'm currently working on a C++ project where I'm using the c-ares library to resolve DNS queries with ares_getaddrinfo. I would like to understand how to effectively correlate the requests I send through c-ares with the responses I receive in my program, particularly in relation to the DNS resolver.

Specifically, I'm looking for guidance on:

1. **Tracking Requests:** What is the best way to associate a request with its corresponding response from the DNS resolver? Are there specific fields or mechanisms within the c-ares API that can help with this?
2. **Handling Multiple Requests:** If I send multiple requests simultaneously, how can I ensure that I can match each response back to its original request?

Any insights or examples would be greatly appreciated!

Thank you!

**Comment by bradh352** [comment]
https://c-ares.org/docs/ares_getaddrinfo.html

Prototype:

> typedef void (*ares_addrinfo_callback)(void *arg, int status,
>                                       int timeouts,
>                                       struct ares_addrinfo *result)
>
> void ares_getaddrinfo(ares_channel_t *channel, const char *name,
>                      const char* service,
>                      const struct ares_addrinfo_hints *hints,
>                      ares_addrinfo_callback callback, void *arg)

Excerpt from docs:

> The callback argument arg is copied from the [ares_getaddrinfo](https://c-ares.org/docs/ares_getaddrinfo.html) argument arg.

So whatever you pass into `arg` when you enqueue the query is passed back to you in the callback's `arg` field, thus providing the correlation you are asking for.

**Reply by arnnav** [reply]
Thanks @bradh352!
If I populate `arg` field, it does provide me correlation between c-ares and my program. But I am not able to correlate my program and the DNS Resolver. I am particularly interested in checking what exactly is happening for a request in the DNS Resolver and what I get returned in my program.

c-ares sends the query to the DNS Resolver with a `query_id`, if I can get this back to my program then I would be able to correlate a request between my program and the DNS Resolver. 
By looking at the code, I can either add this as another parameter in the `ares_getaddrinfo` or I can update the `ares_addrinfo` struct to have the `query_id`. 
What do you think? Any suggestions? 

Basically trying to achieve this, so that I can correlate the logs between DNS Resolver and My Program. 
Using the `query_id` I can see the logs in the DNS Resolver and the response received at my program. 
![image](https://github.com/user-attachments/assets/92e37cc8-2e46-4a80-aa2f-acbfded34b8b)

**Reply by bradh352** [reply]
yeah, right now c-ares doesn't support returning the query id back in ares_getaddrinfo().  You can obviously get it if you use something like ares_send_dnsrec() since you'll get back a structure with the entire DNS message, but then you don't get all the behavior you might want out of ares_getaddrinfo().

Remember, with ares_getaddrinfo() its doing a lot of extra stuff, like 'search' based on ndots and will also send both AF_INET and AF_INET6 queries in parallel if you're using AF_UNSPEC.  So there won't be only a single query id.  There may be a bunch generated for a single request....

**Reply by arnnav** [reply]
Thanks @bradh352.
Would it be beneficial to introduce an additional hint in `ares_addrinfo_hints` struct, in which enabling this hint would result in the `dnsrec` being included in the `ares_addrinfo` struct? This approach would ensure that the changes remain backward compatible when using `ares_getaddrinfo` and we can still get the entire `dnsrec` if needed.

**Reply by arnnav** [reply]
Hi @bradh352, 

I'm considering creating a Pull Request based on the above suggestion (adding a new hint in `ares_addrinfo_hints`). I wanted to get your opinion on whether this would be alright.

Thanks again!
